doc/syntax/pattern_matching.rdoc

```diff
@@ -0,0 +1,364 @@
+
+  jane = {school: 'high', schools: [{id: 1, level: 'middle'}, {id: 2, level: 'high'}]}
+  john = {school: 'high', schools: [{id: 1, level: 'middle'}]}
+
+  case jane
+  in school:, schools: [*, {id:, level: ^school}] # select the last school, level should match
+    "matched. school: #{id}"
+  else
+    "not matched"
+  end
+  #=> "matched. school: 2"
+
+  case john # he specified his school level is "high", but last school does not match
+  in school:, schools: [*, {id:, level: ^school}]
+    "matched. school: #{id}"
+  else
+    "not matched"
+  end
+  #=> "not matched"
+
+== Matching non-primitive objects: +deconstruct_keys+ and +deconstruct+
+
+As already mentioned above, hash and array patterns besides literal arrays and hashes will try to match any object, if it is implementing +deconstruct+ (for array pattern) or +deconstruct_keys+ (for hash pattern).
+
+  class Point
+    def initialize(x, y)
+      @x, @y = x, y
+    end
+
+    def deconstruct
+      puts "deconstruct called"
+      [@x, @y]
+    end
+
+    def deconstruct_keys(keys)
+      puts "deconstruct_keys called with #{keys}"
+      {x: @x, y: @y}
+    end
+  end
+
+  case Point.new(1, -2)
+  in px, Integer  # subpatterns and variable binding works
+    "matched: #{px}"
+  else
+    "not matched"
+  end
+  # prints "deconstruct called"
+  "matched: 1"
+
+  case Point.new(1, -2)
+  in x: 0.. => px
+    "matched: #{px}"
+  else
+    "not matched"
+  end
+  # prints: deconstruct_keys called with [:x]
+  # => "matched: 1"
+
++keys+ are passed to +deconstruct_keys+ to provide room for optimization in matched class: if calculating full hash representation is expensive, it may calculate only the necessary subhash.
+
+Additionally, when matching custom classes, expected class could be specified as part of the pattern and is checked with <code>===</code>
+
+  class SuperPoint < Point
+  end
+
+  case Point.new(1, -2)
+  in SuperPoint(x: 0.. => px)
+    "matched: #{px}"
+  else
+    "not matched"
+  end
+  #=> "not matched"
+
+  case SuperPoint.new(1, -2)
+  in SuperPoint[x: 0.. => px] # [] or () parentheses are allowed
+    "matched: #{px}"
+  else
+    "not matched"
+  end
+  #=> "matched: 1"
+
+== Guard clauses
+
++if+ can be used to attach additional condition (guard clause) when matched pattern
+
+  case [1, 2]
+  in a, b if b == a*2
+    "matched"
+  else
+    "not matched"
+  end
+  #=> "matched"
+
+  case [1, 1]
+  in a, b if b == a*2
+    "matched"
+  else
+    "not matched"
+  end
+  #=> "not matched"
+
++unless+ works, too:
+
+  case [1, 1]
+  in a, b unless b == a*2
+    "matched"
+  else
+    "not matched"
+  end
+  #=> "matched"
+
+== Current feature status
+
+As of Ruby 2.7, feature is considered _experimental_: its syntax can (slightly) change in the future, and the performance is not optimized yet. Every time you use pattern matching in code, the warning will be printed:
```

Comment:
slightly?

Reply:
I was trying to communicate that it is unlikely for the syntax to change completely in upcoming versions, but it is probably inappropriate here. Removed.